File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/A-Минимум_на_стеке/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

class Stack {
private:
    vector<int> stack;
    vector<int> minStack;
public:
    Stack() {};
    
    void push(int x) {
        if (stack.empty()) {
            stack.push_back(x);
            minStack.push_back(x);
        } else {
            stack.push_back(x);
            minStack.push_back(min(minStack.back(), x));
        }
    }
    
    void pop() {
        stack.pop_back();
        minStack.pop_back();
    }
    
    int getMin() {
        return minStack.back();
    }
};


void solution() {
    /*
    Time Complexity O(1) for operations
    Memory Complexity O(N) for storage
    */
    Stack stack = Stack();
    int n;
    scanf("%d", &n);
    for (int i = 0; i < n; ++i) {
        int operation;
        cin >> operation;
        if (operation == 1) {
            int x;
            cin >> x;
            stack.push(x);
        } else if (operation == 2) stack.pop();
        else printf("%d\n", stack.getMin());
    }
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/2sprint-BasicDataStructures/contest/A-Минимум_на_стеке/solution.py
import sys


class Stack:
    def __init__(self) -> None:
        self.stack: list[int] = []
        self.min_stack: list[int] = []
    
    def push(self, x: int) -> None:
        if not self.stack:
            self.stack.append(x)
            self.min_stack.append(x)
        else:
            self.stack.append(x)
            self.min_stack.append(min(self.min_stack[-1], x))
    
    def pop(self) -> None:
        self.stack.pop()
        self.min_stack.pop()
    
    def get_min(self) -> int:
        return self.min_stack[-1]



def solution() -> None:
    """
    Time Complexity O(1) for operations
    Memory Complexity O(N) for storage
    """
    stack: Stack = Stack()
    n: int = int(sys.stdin.readline().rstrip())
    for _ in range(n):
        operation, *args = map(int, sys.stdin.readline().rstrip().split())
        if operation == 1:
            stack.push(args[0])
        elif operation == 2:
            stack.pop()
        else:
            sys.stdout.write(f'{stack.get_min()}\n')


if __name__ == '__main__':
    solution()